Generated policy code fail to compile

```prisma
model Checklist {
    id String @id @default(cuid())
    name String? @default("Checklist")
    creator User @relation(fields: [userId], references: [id])
    activities Activity[]
    userId String
    @@allow('create,read', true)
    @@allow('update,delete', auth() == creator)
}

model Activity {
    id String @id @default(cuid())
    name String
    answers Answer[]
    checklist Checklist @relation(fields: [checklistId], references: [id], onDelete: Cascade)
    checklistId String
    @@allow('create,read', true)
    @@allow('update,delete', auth() == checklist.creator)
}

model Answer {
    id String @id @default(cuid())
    notes String @default("")
    activity Activity @relation(fields: [activityId], references: [id], onDelete: Cascade)
    activityId String
    user User @relation(fields: [userId], references: [id])
    userId String
    @@allow('create', true)
    @@allow('read', auth() == user || auth() == activity.checklist.creator)
    @@allow('read', user == activity.checklist.creator)
    @@allow('update,delete', auth() == user)
}
```

 The `@@allow('read', user == activity.checklist.creator)` line will cause the compiler error
```bash
 error TS2304: Cannot find name 'activity'.

87                 id:((activity?.checklist ?? null)?.creator ?? null).id
                        ~~~~~~~~
```

However change it `@@allow('read', activity.checklist.creator == user)` works

--- a/packages/schema/src/language-server/validator/expression-validator.ts
+++ b/packages/schema/src/language-server/validator/expression-validator.ts
@@ -7,10 +7,12 @@ import {
     isMemberAccessExpr,
     isNullExpr,
     isThisExpr,
+    isDataModelField,
+    isLiteralExpr,
 } from '@zenstackhq/language/ast';
-import { isDataModelFieldReference } from '@zenstackhq/sdk';
+import { isDataModelFieldReference, isEnumFieldReference } from '@zenstackhq/sdk';
 import { ValidationAcceptor } from 'langium';
-import { isAuthInvocation, isCollectionPredicate } from '../../utils/ast-utils';
+import { getContainingDataModel, isAuthInvocation, isCollectionPredicate } from '../../utils/ast-utils';
 import { AstValidator } from '../types';
 import { typeAssignable } from './utils';
 
@@ -125,15 +127,23 @@ export default class ExpressionValidator implements AstValidator<Expression> {
                     accept('error', 'incompatible operand types', { node: expr });
                     break;
                 }
-
                 // not supported:
                 //   - foo.a == bar
                 //   - foo.user.id == userId
-                if(isMemberAccessExpr(expr.left) && isDataModel(expr.left.operand.$resolvedType?.decl) 
-                    || isMemberAccessExpr(expr.right) && isDataModel(expr.right.operand.$resolvedType?.decl))
+                // except:
+                //   - future().userId == userId
+                if(isMemberAccessExpr(expr.left) &&  isDataModelField(expr.left.member.ref) && expr.left.member.ref.$container != getContainingDataModel(expr)
+                || isMemberAccessExpr(expr.right) &&  isDataModelField(expr.right.member.ref) && expr.right.member.ref.$container != getContainingDataModel(expr))
                 {
-                    accept('error', 'comparison between fields of different models are not supported', { node: expr }); 
-                    break;                      
+                    // foo.user.id == auth().id
+                    // foo.user.id == "123"
+                    // foo.user.id == null
+                    // foo.user.id == EnumValue
+                    if(!(this.isNotModelFieldExpr(expr.left) || this.isNotModelFieldExpr(expr.right)))
+                    {
+                        accept('error', 'comparison between fields of different models are not supported', { node: expr }); 
+                        break;  
+                    }                   
                 }
 
                 if (
@@ -194,4 +204,15 @@ export default class ExpressionValidator implements AstValidator<Expression> {
             }
         }
     }
+
+
+    private isNotModelFieldExpr(expr: Expression) {
+        return isLiteralExpr(expr) || isEnumFieldReference(expr) || isNullExpr(expr) || this.isAuthOrAuthMemberAccess(expr)
+    }
+
+    private isAuthOrAuthMemberAccess(expr: Expression) {
+        return isAuthInvocation(expr) || (isMemberAccessExpr(expr) && isAuthInvocation(expr.operand));
+    }
+
 }
+

--- a/packages/schema/src/language-server/validator/expression-validator.ts
+++ b/packages/schema/src/language-server/validator/expression-validator.ts
@@ -13,7 +13,6 @@ import { ValidationAcceptor } from 'langium';
 import { isAuthInvocation, isCollectionPredicate } from '../../utils/ast-utils';
 import { AstValidator } from '../types';
 import { typeAssignable } from './utils';
-import exp from 'constants';
 
 /**
  * Validates expressions.
@@ -127,6 +126,16 @@ export default class ExpressionValidator implements AstValidator<Expression> {
                     break;
                 }
 
+                // not supported:
+                //   - foo.a == bar
+                //   - foo.user.id == userId
+                if(isMemberAccessExpr(expr.left) && isDataModel(expr.left.operand.$resolvedType?.decl) 
+                    || isMemberAccessExpr(expr.right) && isDataModel(expr.right.operand.$resolvedType?.decl))
+                {
+                    accept('error', 'comparison between fields of different models are not supported', { node: expr }); 
+                    break;                      
+                }
+
                 if (
                     (expr.left.$resolvedType?.nullable && isNullExpr(expr.right)) ||
                     (expr.right.$resolvedType?.nullable && isNullExpr(expr.left))
@@ -158,13 +167,6 @@ export default class ExpressionValidator implements AstValidator<Expression> {
                         // incompatible model types
                         // TODO: inheritance case?
                         accept('error', 'incompatible operand types', { node: expr });
-                    }
-
-                    // not supported:
-                    //   - foo.a == bar
-                    if(isMemberAccessExpr(expr.left) || isMemberAccessExpr(expr.left))
-                    {
-                        accept('error', 'comparison between fields of different models are not supported', { node: expr });   
                     }
 
                     // not supported:

--- a/packages/schema/src/language-server/validator/expression-validator.ts
+++ b/packages/schema/src/language-server/validator/expression-validator.ts
@@ -4,6 +4,7 @@ import {
     ExpressionType,
     isDataModel,
     isEnum,
+    isMemberAccessExpr,
     isNullExpr,
     isThisExpr,
 } from '@zenstackhq/language/ast';
@@ -12,6 +13,7 @@ import { ValidationAcceptor } from 'langium';
 import { isAuthInvocation, isCollectionPredicate } from '../../utils/ast-utils';
 import { AstValidator } from '../types';
 import { typeAssignable } from './utils';
+import exp from 'constants';
 
 /**
  * Validates expressions.
@@ -156,6 +158,13 @@ export default class ExpressionValidator implements AstValidator<Expression> {
                         // incompatible model types
                         // TODO: inheritance case?
                         accept('error', 'incompatible operand types', { node: expr });
+                    }
+
+                    // not supported:
+                    //   - foo.a == bar
+                    if(isMemberAccessExpr(expr.left) || isMemberAccessExpr(expr.left))
+                    {
+                        accept('error', 'comparison between fields of different models are not supported', { node: expr });   
                     }
 
                     // not supported:

--- a/packages/schema/src/language-server/zmodel-linker.ts
+++ b/packages/schema/src/language-server/zmodel-linker.ts
@@ -53,7 +53,7 @@ import {
 } from 'langium';
 import { match } from 'ts-pattern';
 import { CancellationToken } from 'vscode-jsonrpc';
-import { getAllDeclarationsFromImports, isAuthInvocation } from '../utils/ast-utils';
+import { getAllDeclarationsFromImports, isAuthInvocation, getContainingDataModel } from '../utils/ast-utils';
 import { mapBuiltinTypeToExpressionType } from './validator/utils';
 
 interface DefaultReference extends Reference {
@@ -292,22 +292,11 @@ export class ZModelLinker extends DefaultLinker {
                 }
             } else if (funcDecl.name === 'future' && isFromStdlib(funcDecl)) {
                 // future() function is resolved to current model
-                node.$resolvedType = { decl: this.getContainingDataModel(node) };
+                node.$resolvedType = { decl: getContainingDataModel(node) };
             } else {
                 this.resolveToDeclaredType(node, funcDecl.returnType);
             }
         }
-    }
-
-    private getContainingDataModel(node: Expression): DataModel | undefined {
-        let curr: AstNode | undefined = node.$container;
-        while (curr) {
-            if (isDataModel(curr)) {
-                return curr;
-            }
-            curr = curr.$container;
-        }
-        return undefined;
     }
 
     private resolveLiteral(node: LiteralExpr) {

--- a/packages/schema/src/language-server/zmodel-linker.ts
+++ b/packages/schema/src/language-server/zmodel-linker.ts
@@ -9,7 +9,6 @@ import {
     DataModelFieldType,
     Enum,
     EnumField,
-    Expression,
     ExpressionType,
     FunctionDecl,
     FunctionParam,

--- a/packages/schema/src/utils/ast-utils.ts
+++ b/packages/schema/src/utils/ast-utils.ts
@@ -156,3 +156,15 @@ export function getAllDeclarationsFromImports(documents: LangiumDocuments, model
 export function isCollectionPredicate(node: AstNode): node is BinaryExpr {
     return isBinaryExpr(node) && ['?', '!', '^'].includes(node.operator);
 }
+
+
+export function getContainingDataModel(node: Expression): DataModel | undefined {
+    let curr: AstNode | undefined = node.$container;
+    while (curr) {
+        if (isDataModel(curr)) {
+            return curr;
+        }
+        curr = curr.$container;
+    }
+    return undefined;
+}

--- a/packages/schema/tests/schema/validation/attribute-validation.test.ts
+++ b/packages/schema/tests/schema/validation/attribute-validation.test.ts
@@ -645,6 +645,35 @@ describe('Attribute tests', () => {
         expect(
             await loadModelWithError(`
             ${prelude}
+            model User {
+                id Int @id
+                lists List[]
+                todos Todo[]
+            }
+              
+            model List {
+                id Int @id
+                user User @relation(fields: [userId], references: [id])
+                userId Int
+                todos Todo[]
+            }
+              
+            model Todo {
+                id Int @id
+                user User @relation(fields: [userId], references: [id])
+                userId Int
+                list List @relation(fields: [listId], references: [id])
+                listId Int
+              
+                @@allow('read', list.user == user)
+            }
+            
+        `)
+        ).toContain('comparison between fields of different models are not supported');
+
+        expect(
+            await loadModelWithError(`
+            ${prelude}
             model A {
                 id String @id
                 x Int

--- a/packages/schema/tests/schema/validation/attribute-validation.test.ts
+++ b/packages/schema/tests/schema/validation/attribute-validation.test.ts
@@ -665,7 +665,7 @@ describe('Attribute tests', () => {
                 list List @relation(fields: [listId], references: [id])
                 listId Int
               
-                @@allow('read', list.user == user)
+                @@allow('read', list.user.id == userId)
             }
             
         `)

